c/misra/src/rules/RULE-21-13/CtypeFunctionArgNotUnsignedCharOrEof.ql

```diff
@@ -0,0 +1,37 @@
+/**
+ * @id c/misra/ctype-function-arg-not-unsigned-char-or-eof
+ * @name RULE-21-13: <ctype.h> function arguments shall be represented as unsigned char
+ * @description Passing arguments to <ctype.h> functions outside the range of unsigned char or EOF
+ *              causes undefined behavior.
+ * @kind problem
+ * @precision very-high
+ * @problem.severity error
+ * @tags external/misra/id/rule-21-13
+ *       external/misra/obligation/mandatory
+ */
+
+import cpp
+import codingstandards.c.misra
+import codingstandards.cpp.ReadErrorsAndEOF
+import semmle.code.cpp.rangeanalysis.SimpleRangeAnalysis
+
+class CtypeFunction extends Function {
+  CtypeFunction() { this.getADeclaration().getAFile().(HeaderFile).getBaseName() = "ctype.h" }
+}
+
+from FunctionCall ctypeCall
```

Comment:
I have created a new `UseOfToOrIsChar` class:

https://github.com/github/codeql-coding-standards/blob/main/cpp/common/src/codingstandards/cpp/CharFunctions.qll#L20

That can be used to abstract away the difference between whether the "function" is implemented with a real function or with a macro.

You can see an example of how to use the class here:
https://github.com/github/codeql-coding-standards/blob/466eae878fbc86250e749cc5c0b072abdb78af69/c/cert/src/rules/STR37-C/ToCharacterHandlingFunctionsRepresentableAsUChar.ql#L19

You can use `getConvertedArgument()` to get the argument for which you can check the lower bound/upper bound.

Reply:
Completed.